python/Scripts/genmdl.py

```diff
@@ -0,0 +1,217 @@
+    file.write('import::base::*;\n')
+    file.write('import::df::*;\n')
+    file.write('import::math::*;\n')
+    file.write('import::state::*;\n')
+    file.write('import::tex::*; \n')                        
+    file.write('\n')
+
+def main():
+
+    if len(sys.argv) < 2:
+        usage()
+        sys.exit(0)
+
+    _startPath = os.path.abspath(sys.argv[1])
+    if os.path.exists(_startPath) == False:
+        print('Start path does not exist: ' + _startPath + '. Using current directory.\n')
+        _startPath = os.path.abspath(os.getcwd())
+
+    moduleName = ''
+    if len(sys.argv) > 2:
+        moduleName = sys.argv[2]
+
+    LIBRARY = 'stdlib'
+
+    doc = mx.createDocument()
+    searchPath = os.path.join(_startPath, 'libraries')
+    libraryPath = os.path.join(searchPath, 'stdlib')
+    _loadLibraries(doc, searchPath, libraryPath)
+
+    DEFINITION_PREFIX = 'ND_'
+    IMPLEMENTATION_PREFIX = 'IM_'
+    IMPLEMENTATION_STRING = 'impl'
+    GENMDL = 'genmdl'
+
+    # Create target directory if don't exist
+    outputPath = os.path.join(libraryPath, GENMDL)
+    if not os.path.exists(outputPath):
+        os.mkdir(outputPath)
+
+    file = None
+    
+    # Write to single file if module name specified
+    if len(moduleName):
+        file = open(outputPath + '/' + moduleName + '.mdl', 'w+')
+        _writeImports(file)
+        
+    # Dictionary to map from MaterialX type declarations 
+    # to MDL type declarations
+    typeMap = dict()
+    typeMap['boolean'] = 'bool'
+    typeMap['integer'] = 'int'
+    typeMap['color2'] = 'float2'
+    typeMap['color4'] = 'float4'
+    typeMap['vector2'] = 'float2'
+    typeMap['vector3'] = 'float3'
+    typeMap['vector4'] = 'float4'
+    typeMap['matrix33'] = 'float3x3'
+    typeMap['matrix44'] = 'float4x4'
+    typeMap['filename'] = 'string'
+    typeMap['geomname'] = 'string'
+    typeMap['floatarray'] = 'float[<count>]'
+    typeMap['integerarray'] = 'int[<count>]'
+    typeMap['color2array'] = 'float2[<count>]'
+    typeMap['color3array'] = 'color[<count>]'
+    typeMap['color4array'] = 'float4[<count>]'
+    typeMap['vector2array'] = 'float2[<count>]'
+    typeMap['vector3array'] = 'float3[<count>]'
+    typeMap['vector4array'] = 'float4[<count>]'
+    typeMap['stringarray'] = 'string[<count>]'
+    typeMap['geomnamearray'] = 'string[<count>]'
+
+    INDENT = '  '
+
+    # Create an implementation per nodedef
+    #
+    implDoc = mx.createDocument()
+    nodedefs = doc.getNodeDefs()
+    for nodedef in nodedefs:
+
+        if len(nodedef.getActiveOutputs()) == 0:
+           continue
+
+        # String out definition prefix
+        nodeName = nodedef.getName()
+        if len(nodeName) > 3:
+            if (nodeName[0:3] == DEFINITION_PREFIX):
+                nodeName = nodeName[3:]
+
+        filename = nodeName + '.mdl'
+
+        implname = IMPLEMENTATION_PREFIX + nodeName + '_' + GENMDL
+        impl = implDoc.addImplementation(implname)                    
```

Comment:
This could be very handy to generate the implementation elements, especially if it turns out we will use inline expressions for all (as I suspect). Then we could update this to emit the inline expression as well into an attribute.

So maybe merge in his script but leave the two generated files outside for now?